lib/lita/adapters/slack/rtm_connection.rb

```diff
@@ -46,29 +46,43 @@ def im_for(user_id)
           im_mapping.im_for(user_id)
         end
 
-        def run(queue = nil, options = {})
-          EventLoop.run do
-            log.debug("Connecting to the Slack Real Time Messaging API.")
-            @websocket = Faye::WebSocket::Client.new(
-              websocket_url,
-              nil,
-              websocket_options.merge(options)
-            )
-
-            websocket.on(:open) do
-              log.debug("Connected to the Slack Real Time Messaging API.")
-              yield if block_given?
-            end
-            websocket.on(:message) { |event| receive_message(event) }
-            websocket.on(:close) do |event|
-              log.info("Disconnected from Slack.")
-              log.info(event.code)
-              log.info(event.reason)
+        def create_websocket(queue, options, &block)
+          ws = Faye::WebSocket::Client.new(
+            websocket_url,
+            nil,
+            websocket_options.merge(options)
+          )
+
+          ws.on(:open) do
+            log.debug("Connected to the Slack Real Time Messaging API.")
+            yield if block_given?
+          end
+          ws.on(:message) { |event| receive_message(event) }
+          ws.on(:close) do |event|
+            log.info("Disconnected from Slack: #{event.code} - #{event.reason}")
+
+            # Slack periodically closes the websocket connection
+            if event.code.to_s == '1006'
+              queue.delete(@websocket) if queue
+              @team_data = API.new(@config).rtm_start
+              @im_mapping = IMMapping.new(API.new(@config), @team_data.ims)
+              @websocket_url = @team_data.websocket_url
+              @robot_id = @team_data.self.id
+              @websocket = create_websocket(queue, options, &block)
```

Comment:
Why do the instance var's above need to be updated?

Reply:
The initial attempt used to only set `@websocket`, but that didn't refresh the websocket URL; we didn't even know that refreshing that was _a thing_.

We could try setting up just `@websocket_url` and `@websocket`, but in the end we opted for replicating the startup process as much as possible here - this _should_ cover us up if there's any other unknown fields that should be refreshed.